routing/index_router.cpp

```diff
@@ -558,29 +592,39 @@ bool IndexRouter::FindBestSegment(m2::PointD const & point, m2::PointD const & d
   vector<pair<Edge, Junction>> candidates;
   m_roadGraph.FindClosestEdges(point, kMaxRoadCandidates, candidates);
 
-  auto const getSegmentByEdge = [&numMwmId](Edge const & edge){
-      return Segment(numMwmId, edge.GetFeatureId().m_index, edge.GetSegId(), edge.IsForward());
+  auto const getSegmentByEdge = [&numMwmId](Edge const & edge, bool isForward) {
+    return Segment(numMwmId, edge.GetFeatureId().m_index, edge.GetSegId(), isForward);
   };
 
   // Getting rid of knowingly bad candidates.
   my::EraseIf(candidates, [&](pair<Edge, Junction> const & p){
     Edge const & edge = p.first;
-    return edge.GetFeatureId().m_mwmId != mwmId || IsDeadEnd(getSegmentByEdge(edge), isOutgoing, worldGraph);
+    return edge.GetFeatureId().m_mwmId != mwmId ||
+           IsDeadEnd(getSegmentByEdge(edge, edge.IsForward()), isOutgoing, worldGraph);
   });
 
   if (candidates.empty())
     return false;
 
-  BestEdgeComparator bestEdgeComparator(point, direction);
-  Edge & bestEdge = candidates[0].first;
+  BestEdgeComparator bestEdgeComparator(point, direction, [this](FeatureID const & featureId) {
+    return this->IsOneWayFeature(featureId);
+  });
+  Edge bestEdge = candidates[0].first;
   for (size_t i = 1; i < candidates.size(); ++i)
   {
     Edge const & edge = candidates[i].first;
     if (bestEdgeComparator.Compare(edge, bestEdge))
       bestEdge = edge;
   }
 
-  bestSegment = getSegmentByEdge(bestEdge);
+  bestSegmentIsAlmostConformDirection =
+      bestEdgeComparator.IsDirectionValid() && bestEdgeComparator.IsAlmostConformed(bestEdge);
+
+  bool bestSegmentForward = true;
+  if (bestEdgeComparator.IsDirectionValid())
+    bestEdgeComparator.IsAlmostCollinear(bestEdge, bestSegmentForward);
```

Comment:
Yes're right. The code here isn't clear enough. What's up here (see new comments below):
```
bool IndexRouter::FindBestSegment(m2::PointD const & point, m2::PointD const & direction,
                                  bool isOutgoing, WorldGraph & worldGraph, Segment & bestSegment,
                                  bool & bestSegmentIsAlmostConformDirection) const
{
.....
// 1. Looking for the best edge according to |bestEdgeComparator|.
  Edge bestEdge = candidates[0].first;
  for (size_t i = 1; i < candidates.size(); ++i)
  {
    Edge const & edge = candidates[i].first;
    if (bestEdgeComparator.Compare(edge, bestEdge))
      bestEdge = edge;
  }

// Checking if the best edge is taken because of its angle.
// bestSegmentIsAlmostConformDirection == true if so. 
// bestSegmentIsAlmostConformDirection == false there's no edge in neighbourhoods which have angle close to |BestEdgeComparator::m_direction|. It means |bestEdge| is chosen as closest edge to |point| and its direction(forward == true) wont be used.
  bestSegmentIsAlmostConformDirection =
      bestEdgeComparator.IsDirectionValid() && bestEdgeComparator.IsAlmostConformed(bestEdge);

// Getting |m_forward| member for |bestSegment|.
// The call bestEdgeComparator.IsAlmostCollinear(bestEdge, bestSegmentForward) fills |bestSegmentForward|. There are the following option: 
 * |bestSegmentIsAlmostConformDirection| == false (|bestEdge| is chosen by |point|)=>|bestSegmentForward|/Segment::m_forward will be true.
 * |bestSegmentIsAlmostConformDirection| == true and |bestSegmentForward|/Segment::m_forward will be true if vectors |BestEdgeComparator::m_direction| and Segment are almost parallel and false otherwise.
  bool bestSegmentForward = true;
  if (bestEdgeComparator.IsDirectionValid())
    bestEdgeComparator.IsAlmostCollinear(bestEdge, bestSegmentForward);

  bestSegment = getSegmentByEdge(bestEdge, bestSegmentForward);
  return true;
}
```

The main problem I see with the code it's not difficult for understanding. I'll think if it's possible to refactor it to make it more clear. Any ideas are welcomed:)